gcc/rust/resolve/rust-rib.h

```diff
@@ -0,0 +1,133 @@
+#define RUST_RIB_H
+
+#include "rust-system.h"
+#include "rust-ast.h"
+#include "optional.h"
+#include "expected.h"
+
+namespace Rust {
+namespace Resolver2_0 {
+
+/**
+ * All namespaces that Rust's name resolution needs to handle
+ */
+// TODO: Move to `rust-forever-stack.h`?
+enum class Namespace
+{
+  Values,
+  Types,
+  Labels,
+  Macros,
+  // TODO: Which namespaces are we missing?
+};
+
+/**
+ * Error returned by `Rib::insert` when the key was already present in the Rib's
+ * map. The class contains the previously-inserted NodeId as well as the name of
+ * the node.
+ */
+struct AlreadyInserted
+{
+  // TODO: We might need multiple kinds of errors later down the line
+  AlreadyInserted (Identifier name, NodeId existing);
+
+  Identifier name;
+  NodeId existing;
+};
+
+/**
+ * A rib is a container of nodes, either declaration or usages, as well as the
+ * identifier each node uses. They are used to delimit lexical scopes, and have
+ * an impact on name resolution - they restrict certain name accesses and serve
+ * as boundaries between scopes.
+
+ * For example, if we are resolving the following *variable* use:
+ *
+ * ```rust
+ * fn outer() {
+ *     let a = 15; // decl
+ *     fn inner() -> i32 {
+ *         a // use
+ *     }
+ * }
+ * ```
+ *
+ * The `Function` rib we will have pushed will restrict the access to `outer`'s
+ * `a` declaration: Variable uses cannot cross function boundaries. On the other
+ * hand, if we were resolving a type usage, this would be perfectly allowed.
+ */
+class Rib
+{
+public:
+  enum class Kind
+  {
+    Normal,
+    Module,
+    Function,
+    Item, // TODO: what's an item scope? Is that the original crate's scope?
+    TraitOrImpl,
+    Closure,
+    ConstantItem, // -> this variant has a boolean
+    MacroDefinition,
+    ForwardTypeParam, // TODO: What is this one for?
+    ConstParamType,   // TODO: Const parameter ? Like... Constant item? Or Const
+		      // generic? */
+  };
+
+  Rib (Kind kind);
+  Rib (Kind kind, Identifier identifier, NodeId id);
+  Rib (Kind kind, std::unordered_map<Identifier, NodeId> values);
+
+  // TODO: What's the correctbehavior if the key already exists? What if a decl
+  // and use are in the same rib? Is that possible? Okay based on RibKind?
+
+  /**
+   * Insert a new node in the rib
+   *
+   * @param name The name associated with the AST node
+   * @param id Its NodeId
+   *
+   * @return `AlreadyInserted` if the node is already present in the rib. The
+   *         `AlreadyInserted` class contains the NodeId of the existing node.
+   *         Returns the new NodeId on success.
+   */
+  tl::expected<NodeId, AlreadyInserted> insert (Identifier name, NodeId id);
+
+  /**
+   * Access an inserted NodeId.
+   *
+   * @return tl::nullopt if the key does not exist, the NodeId otherwise
+   */
+  tl::optional<NodeId> get (const Identifier &name);
+
+  /* View all the values stored in the rib */
+  const std::unordered_map<Identifier, NodeId> &get_values () const;
```

Comment:
Good idea to start using Identifier but we need to make sure it has a comparator function

Reply:
This set should only be "name-based" so I guess maybe it would be better to store the Identifier's string as a key with the new changes being done to that class. Because the `values` unordered_map contains the Identifier's NodeId as a value. Or maybe it would be better to turn this into an `std::set<Identifier>`. I don't know

Reply:
So for this set I think we should go with an `unordered_map<string, NodeId>`. Let's say we're inserting a new function in the rib: we care about the function's name, so we can use its Identifier, but we care about the function's node ID - not the identifier's node ID. So I don't think having a `set<Identifier>` makes sense for the rib. The one place where we will be using the Identifier's NodeId for resolution is for resolving proc macros, as @P-E-P pointed it out